kotlin/samples/hello-terminal/terminal-workflow/src/main/java/com/squareup/sample/helloterminal/terminalworkflow/TerminalSize.kt

```diff
@@ -36,18 +36,25 @@ data class TerminalSize(
   val columns: Int
 )
 
+/**
+ * A [Flow] of the current terminal size. Will always emit the current size immediately on
+ * collection.
+ */
 @UseExperimental(ExperimentalCoroutinesApi::class)
-internal fun Terminal.listenForResizesOn(
-  scope: CoroutineScope
-): ReceiveChannel<TerminalSize> =
-// Run with unconfined dispatcher because this is just sending events to a channel, we don't care
-  // what thread it's on.
-  scope.produce(context = Unconfined, capacity = Channel.CONFLATED) {
+internal val Terminal.size: Flow<TerminalSize>
+  get() = channelFlow<TerminalSize> {
     val resizeListener = TerminalResizeListener { _, newSize -> offer(newSize.toSize()) }
-    invokeOnClose { removeResizeListener(resizeListener) }
     addResizeListener(resizeListener)
-    // Suspend until cancelled.
-    suspendCancellableCoroutine<Nothing> { }
+    awaitClose { removeResizeListener(resizeListener) }
   }
+      // Conflate first, so it fuses with the channelFlow operator.
+      .conflate()
+      .startWith(terminalSize.toSize())
 
 internal fun LanternaTerminalSize.toSize(): TerminalSize = TerminalSize(rows, columns)
+
+@UseExperimental(ExperimentalCoroutinesApi::class)
+private fun <T> Flow<T>.startWith(value: T): Flow<T> = flow {
```

Comment:
Had to write a custom `startWith` as well, but this is super trivial and if we merge this I will submit a PR to add it to the coroutines library.

Reply:
This implementation is broken. The upstream flow won't get collected until the first emit returns. If the upstream flow is hot (as it is for resizes) and an event occurs while the first emit is suspended, those events will be dropped. Needs to open a channel to the upstream before emitting the initial value.

Reply:
Are you going to document that or fix it?

Reply:
Fix